src/Symfony/Component/VarDumper/Cloner/VarCloner.php

```diff
@@ -41,92 +39,87 @@ protected function doClone($var)
-            $j = -1;                    // Position in the currently iterated array
-            $fromObjCast = array_keys($queue[$i]);
-            $fromObjCast = array_keys(array_flip($fromObjCast)) !== $fromObjCast;
-            $refs = $vals = $fromObjCast ? array_values($queue[$i]) : $queue[$i];
-            foreach ($queue[$i] as $k => $v) {
-                // $k is the original key
+            $fromObjCast = \array_keys($queue[$i]);
+            $fromObjCast = \array_keys(\array_flip($fromObjCast)) !== $fromObjCast;
+            $refs = $vals = $fromObjCast ? \array_values($queue[$i]) : $queue[$i];
+            foreach ($vals as $k => $v) {
                 // $v is the original value or a stub object in case of hard references
-                if ($k !== ++$j) {
-                    $indexed = false;
-                }
-                if ($fromObjCast) {
-                    $k = $j;
-                }
-                if ($useExt) {
-                    $zval = symfony_zval_info($k, $refs);
-                } else {
-                    $refs[$k] = $cookie;
-                    if ($zval['zval_isref'] = $vals[$k] === $cookie) {
-                        $zval['zval_hash'] = $v instanceof Stub ? spl_object_hash($v) : null;
-                    }
-                    $zval['type'] = gettype($v);
-                }
-                if ($zval['zval_isref']) {
+                $refs[$k] = $cookie;
+                if ($zvalIsRef = $vals[$k] === $cookie) {
                     $vals[$k] = &$stub;         // Break hard references to make $queue completely
                     unset($stub);               // independent from the original structure
-                    if (isset($hardRefs[$zval['zval_hash']])) {
-                        $vals[$k] = $useExt ? ($v = $hardRefs[$zval['zval_hash']]) : ($refs[$k] = $v);
+                    if (isset($hardRefs[$v instanceof Stub ? \spl_object_hash($v) : ''])) {
+                        $vals[$k] = $refs[$k] = $v;
                         if ($v->value instanceof Stub && (Stub::TYPE_OBJECT === $v->value->type || Stub::TYPE_RESOURCE === $v->value->type)) {
                             ++$v->value->refCount;
                         }
                         ++$v->refCount;
                         continue;
                     }
+                    $refs[$k] = $vals[$k] = new Stub();
+                    $refs[$k]->value = $v;
+                    $h = \spl_object_hash($refs[$k]);
+                    $hardRefs[$h] = &$refs[$k];
+                    $values[$h] = $v;
+                    $vals[$k]->handle = ++$refsCounter;
                 }
                 // Create $stub when the original value $v can not be used directly
                 // If $v is a nested structure, put that structure in array $a
-                switch ($zval['type']) {
-                    case 'string':
-                        if (isset($v[0]) && !preg_match('//u', $v)) {
+                switch (true) {
```

Comment:
Using root namespaced `\is_*()` function is noticeably faster that comparing the string result of `gettype()` on PHP 7.
Since this is a hot path, I've added a `\` to all root functions in the loop.